src/hotspot/share/opto/c2_globals.hpp

```diff
@@ -470,6 +470,12 @@
   develop(bool, TracePostallocExpand, false, "Trace expanding nodes after"  \
           " register allocation.")                                          \
                                                                             \
+  product(bool, ReduceAllocationMerges, true,                               \
```

Comment:
You can look at it in the following way: since the flag is set to true by default, the feature is unconditionally available in product binaries. The only reason to explicitly specify the flag is to turn the optimization off and it may be needed to diagnose VM crashes or performance regressions. 

So, maybe C2 should check a compiler directive (and not a global flag) to be able to control the optimization up to per-method granularity.